crypto/bio/bio_mem.c

```diff
@@ -98,56 +105,87 @@ BIO *BIO_new_mem_buf(const void *buf, ossl_ssize_t len) {
 }
 
 static int mem_new(BIO *bio) {
-  BUF_MEM *b;
+  BIO_BUF_MEM *bbm = OPENSSL_zalloc(sizeof(*bbm));
+
+  if (bbm == NULL) {
+    return 0;
+  }
 
-  b = BUF_MEM_new();
-  if (b == NULL) {
+  bbm->buf = BUF_MEM_new();
+  if (bbm->buf == NULL) {
+    OPENSSL_free(bbm);
     return 0;
   }
 
   // |shutdown| is used to store the close flag: whether the BIO has ownership
   // of the BUF_MEM.
+  bbm->off = 0;
   bio->shutdown = 1;
   bio->init = 1;
   bio->num = -1;
-  bio->ptr = (char *)b;
+  bio->ptr = (char *)bbm;
 
   return 1;
 }
 
 static int mem_free(BIO *bio) {
+  BIO_BUF_MEM *bbm;
+
+  bbm = (BIO_BUF_MEM *)bio->ptr;
+  BUF_MEM *b = bbm->buf;
+
   if (!bio->shutdown || !bio->init || bio->ptr == NULL) {
     return 1;
   }
 
-  BUF_MEM *b = (BUF_MEM *)bio->ptr;
   if (bio->flags & BIO_FLAGS_MEM_RDONLY) {
     b->data = NULL;
   }
   BUF_MEM_free(b);
   bio->ptr = NULL;
+
+  OPENSSL_free(bbm);
   return 1;
 }
 
+static int mem_buf_sync(BIO *bio) {
+  if (bio != NULL && bio->init != 0 && bio->ptr != NULL) {
+    BIO_BUF_MEM *bbm = (BIO_BUF_MEM *) bio->ptr;
+    BUF_MEM *b = bbm->buf;
+
+    if (b->data != NULL) {
+      OPENSSL_memmove(b->data, &b->data[bbm->off], b->length);
+      bbm->off = 0;
+    }
+  }
+
+  return 0;
+}
+
 static int mem_read(BIO *bio, char *out, int outl) {
   BIO_clear_retry_flags(bio);
   if (outl <= 0) {
     return 0;
   }
 
-  BUF_MEM *b = bio->ptr;
+  BIO_BUF_MEM *bbm = (BIO_BUF_MEM *) bio->ptr;
+  BUF_MEM *b = bbm->buf;
+
   int ret = outl;
   if ((size_t)ret > b->length) {
     ret = (int)b->length;
   }
 
   if (ret > 0) {
-    OPENSSL_memcpy(out, b->data, ret);
+    size_t off = bbm->off;
+    if (bio->flags & BIO_FLAGS_MEM_RDONLY) {
+      off = 0;
+    }
+    OPENSSL_memcpy(out, &b->data[off], ret);
```

Comment:
would it be simpler if we just always use `bbm->off` on L184 and eliminate the `b->data += ret` advancement on L188 as well as other manipulation of `b->data`?

it seems a bit odd that we modify the value of `b->data`, but only in r/o mode.

if not, suggest adding a comment why you have to override `off = 0` in the r/o case

Reply:
This behavior of modifying `b->data` in R/O mode is pretty consistent in both BoringSSL and OpenSSL. 

In R/O mode, the getter functions (`BIO_get_mem_ptr`, `BIO_get_mem_data`, and `BIO_mem_contents`) expect that the modified `b->data` to be returned. This is especially important for `BIO_get_mem_ptr` that's returning the entire `BUF_MEM` struct, not just `b->data`. Therefore, I decided to keep `b->data += ret` for better backward compatibility.

Reply:
Added comments + refactored the code a little to make it clear that `read_off` is to keep track of the number of bytes already read from the original buffer, not necessarily the relative offset from the current `b->data` ptr

Reply:
Nvm! Talked to Justin and refactored the code to simplify the way we handle the memory buffer in R/O mode